test/e2e/scheduling/predicates.go

```diff
@@ -724,6 +727,231 @@ var _ = SIGDescribe("SchedulerPredicates [Serial]", func() {
+			pollInterval := 1 * time.Second
+			err = testutils.WaitForDeploymentCompleteAndCheckRolling(cs, d, framework.Logf, pollInterval, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled", deploymentName)
+			err = wait.Poll(pollInterval, wait.ForeverTestTimeout, deploymentReplicasScheduled(ctx, cs, ns, deploymentPodLabels))
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in deployment %q is scheduled on the same Node", deploymentName)
+			nodenames, err := deploymentReplicasNodeNames(ctx, cs, ns, deploymentPodLabels)
+			framework.ExpectNoError(err)
+			nodename := nodenames[0]
+			for _, n := range nodenames {
+				framework.ExpectEqual(nodename, n)
+			}
+
+			framework.Logf("update deployment %q to run rolling upgrade", deploymentName)
+			// change label to kick the rolling upgrade.
+			d.Spec.Template.Labels["test"] = "matchLabelKey2"
+			d, err = cs.AppsV1().Deployments(ns).Update(ctx, newDeployment, metav1.UpdateOptions{})
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure updated deployment %q is complete", deploymentName)
+			err = testutils.WaitForDeploymentCompleteAndCheckRolling(cs, d, framework.Logf, pollInterval, wait.ForeverTestTimeout)
+			framework.ExpectNoError(err)
+
+			framework.Logf("Make sure all Pods in updated deployment %q is scheduled", deploymentName)
+			err = wait.Poll(pollInterval, wait.ForeverTestTimeout, deploymentReplicasScheduled(ctx, cs, ns, deploymentPodLabels))
+			framework.ExpectNoError(err)
+
+			// new replicas shouldn't be affected by existing replicas.
+			framework.Logf("Make sure all Pods in updated deployment %q is scheduled on the same Node", deploymentName)
+			nodenames, err = deploymentReplicasNodeNames(ctx, cs, ns, deploymentPodLabels)
+			framework.ExpectNoError(err)
+			nodename2 := nodenames[0]
+			for _, n := range nodenames {
+				framework.ExpectEqual(nodename2, n)
+			}
+
+			// The second deployment should be scheduled on a Node different from the first deployment's Pods.
+			framework.ExpectNotEqual(nodename2, nodename)
+		})
+
+		ginkgo.It("PodAntiAffinity: Pods from the same replicaset are taken into account by matchLabelKeys during the deployment's rolling upgrade [Feature:MatchLabelKeysInPodAffinity] [alpha]", func(ctx context.Context) {
+			// It's a real-world usecase of matchLabelKeys.
+			// Create a deployment, run rolling upgrade, and see new replicas are spread over all Nodes.
+			// The replica num will be set to be the same as the number of Node.
+			// So, if the matchLabelKeys feature is broken, no new Pod can be scheduled during the rolling upgrade and this test fails.
+
+			e2eskipper.SkipUnlessNodeCountIsAtLeast(3)
+			numNode := framework.TestContext.CloudConfig.NumNodes
+			ginkgo.By("make the nodes have balanced resource usage ratio")
+			err := createBalancedPodForNodes(ctx, f, cs, ns, nodeList.Items, podRequestedResource, 0.5)
```

Comment:
this is an expensive operation. Do you really need it? or, you can just use lightweight solutions to achieve the same goal? (applies to PodAffinity e2e as well)

Reply:
I took a look at the e2e logic, and seems it doesn't really need to create balanced pods out. Did I miss anything?

Reply:
I _was_ thinking there are some Pods left from the previous test cases. Seems it's not true.
I'll remove this line